File: C-SharpTest/C-SharpTest/OpenGloveAPI_C_Sharp_HL/Communication.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using WebSocketSharp; //Nuget: WebSocketSharp-netstandard (1.0.1)

namespace CSharpTest.OpenGloveAPI_C_Sharp_HL
{
    public class Communication
    {
        public WebSocket WebSocket { get; set; }
        public MessageGenerator MessageGenerator { get; set; }
        public string BluetoothDeviceName { get; set; }
        public string ConfigurationName { get; set; }

        public delegate void TimeTestServerLatencyActivateActuatorsReceived(long nanoSeconds);
        public delegate void TimeTestArduinoLatencyActivateActuatorsReceived(long microSeconds);
        public delegate void FlexorMovement(int region, int value);
        public delegate void AccelerometerValues(float ax, float ay, float az);
        public delegate void GyroscopeValues(float gx, float gy, float gz);
        public delegate void MagnometerValues(float mx, float my, float mz);
        public delegate void AttitudeValues(float pitch, float roll, float yaw);
        public delegate void AllIMUValues(float ax, float ay, float az, float gx, float gy, float gz, float mx, float my, float mz);
        public delegate void BluetoothDeviceConnectionState(bool isConnected);
        public delegate void WebSocketConnectionState(bool isConnected);
        public delegate void InfoMessage(string message);

        public event TimeTestServerLatencyActivateActuatorsReceived OnTimeTestServerLatencyActivateActuatorsReceived;
        public event TimeTestArduinoLatencyActivateActuatorsReceived OnTimeTestArduinoLatencyActivateActuatorsReceived;
        public event FlexorMovement OnFlexorValueReceived;
        public event AccelerometerValues OnAccelerometerValuesReceived;
        public event GyroscopeValues OnGyroscopeValuesReceived;
        public event MagnometerValues OnMagnometerValuesReceived;
        public event AttitudeValues OnAttitudeValuesReceived;
        public event AllIMUValues OnAllIMUValuesReceived;
        public event BluetoothDeviceConnectionState OnBluetoothDeviceConnectionStateChanged;
        public event WebSocketConnectionState OnWebSocketConnectionStateChangued;
        public event InfoMessage OnInfoMessagesReceived;

        public Communication(string bluetoothDeviceName, string configurationName, string url)
        {
            this.WebSocket = new WebSocket(url);
            this.WebSocket.OnOpen += OnOpen;         
            this.WebSocket.OnMessage += OnMessage;
            this.WebSocket.OnClose += OnClose;
            this.WebSocket.OnError += OnError;
            this.BluetoothDeviceName = bluetoothDeviceName;
            this.ConfigurationName = configurationName;

            MessageGenerator = new MessageGenerator(mainSeparator: ";", secondarySeparator: ",", empty: "");
        }

        private void OnOpen(object sender, EventArgs e)
        {
            Debug.WriteLine($"Websocket Connected to Server!");
            this.AddOpenGloveDeviceToServer(BluetoothDeviceName, ConfigurationName);
            Debug.WriteLine($"OpenGlove.Communication.AddOpenGloveToServer()");
            this.StartCaptureDataFromServer(BluetoothDeviceName);
            Debug.WriteLine($"Communication.StartCaptureDataFromServer()");
            OnWebSocketConnectionStateChangued?.Invoke(this.WebSocket.IsAlive);

        }

        private void OnMessage(object sender, MessageEventArgs e)
        {
            Debug.WriteLine($"Received from Server: {e.Data}");
            MessageHandler(e.Data);
        }

        private void MessageHandler(string message)
        {
            int mapping, value;
            float valueX, valueY, valueZ, pitch, roll, yaw;
            string[] words;

            if (message != null)
            {
                words = message.Split(',');
                try
                {
                    switch (words[0])
                    {
                        case "f":
                            mapping = Int32.Parse(words[1]);
                            value = Int32.Parse(words[2]);
                            OnFlexorValueReceived?.Invoke(mapping, value);
                            break;
                        case "a":
                            valueX = float.Parse(words[1], CultureInfo.InvariantCulture);
                            valueY = float.Parse(words[2], CultureInfo.InvariantCulture);
                            valueZ = float.Parse(words[3], CultureInfo.InvariantCulture);
                            OnAccelerometerValuesReceived?.Invoke(valueX, valueY, valueZ);
                            break;
                        case "g":
                            valueX = float.Parse(words[1], CultureInfo.InvariantCulture);
                            valueY = float.Parse(words[2], CultureInfo.InvariantCulture);
                            valueZ = float.Parse(words[3], CultureInfo.InvariantCulture);
                            OnGyroscopeValuesReceived?.Invoke(valueX, valueY, valueZ);
                            break;
                        case "m":
                            valueX = float.Parse(words[1], CultureInfo.InvariantCulture);
                            valueY = float.Parse(words[2], CultureInfo.InvariantCulture);
                            valueZ = float.Parse(words[3], CultureInfo.InvariantCulture);
                            OnMagnometerValuesReceived?.Invoke(valueX, valueY, valueZ);
                            break;
                        case "z":
                            OnAllIMUValuesReceived?.Invoke(float.Parse(words[1], CultureInfo.InvariantCulture), float.Parse(words[2], CultureInfo.InvariantCulture), float.Parse(words[3], CultureInfo.InvariantCulture), float.Parse(words[4], CultureInfo.InvariantCulture), float.Parse(words[5], CultureInfo.InvariantCulture), float.Parse(words[6], CultureInfo.InvariantCulture), float.Parse(words[7], CultureInfo.InvariantCulture), float.Parse(words[8], CultureInfo.InvariantCulture), float.Parse(words[9], CultureInfo.InvariantCulture));
                            break;
                        case "r":
                            pitch = float.Parse(words[1], CultureInfo.InvariantCulture);
                            roll = float.Parse(words[2], CultureInfo.InvariantCulture);
                            yaw = float.Parse(words[3], CultureInfo.InvariantCulture);
                            OnAttitudeValuesReceived?.Invoke(pitch, roll, yaw);
                            break;
                        case "b":
                            OnBluetoothDeviceConnectionStateChanged?.Invoke(bool.Parse(words[1]));
                            break;
                        case "us":
                            OnTimeTestArduinoLatencyActivateActuatorsReceived?.Invoke(long.Parse(words[1]));
                            break;
                        case "ns":
                            OnTimeTestServerLatencyActivateActuatorsReceived?.Invoke(long.Parse(words[1]));
                            break;
                            
                        default:
                            OnInfoMessagesReceived?.Invoke(message);
                            break;
                    }

                }
                catch
                {
                    Debug.WriteLine($"ERROR: BAD FORMAT on Communication.MessageHandler [message: {message}]");
                }
            }
        }

        private void OnClose(object sender, EventArgs e)
        {
            Debug.WriteLine($"WebSocket Closed: {e.ToString()}");
        }

        private void OnError(object sender, ErrorEventArgs e)
        {
            Debug.WriteLine($"WebSocket Error: {e.Exception}, {e.ToString()}");
        }

        public void StartOpenGlove(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.StartOpenGlove(bluetoothDeviceName));
        }

        public void StopOpenGlove(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.StopOpenGlove(bluetoothDeviceName));
        }

        public void AddOpenGloveDeviceToServer(string bluetoothDeviceName, string configurationName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.AddOpenGloveDeviceToServer(bluetoothDeviceName, configurationName));
        }

        public void RemoveOpenGloveDeviceFromServer(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.RemoveOpenGloveDeviceFromServer(bluetoothDeviceName));
        }
        public void SaveOpenGloveConfiguration(string bluetoothDeviceName, string configurationName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.SaveOpenGloveConfiguration(bluetoothDeviceName, configurationName));
        }

        public void ConnectToBluetoothDevice(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.ConnectToBluetoothDevice(bluetoothDeviceName));
        }

        public void DisconnectFromBluetoothDevice(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.DisconnectFromBluetoothDevice(bluetoothDeviceName));
        }

        public void StartCaptureDataFromServer(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.StartCaptureDataFromServer(bluetoothDeviceName));
        }

        public void StopCaptureDataFromServer(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.StopCaptureDataFromServer(bluetoothDeviceName));
        }

        public void AddActuator(string bluetoothDeviceName, int region, int positivePin, int negativePin)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.AddActuator(bluetoothDeviceName, region, positivePin, negativePin));
        }

        public void AddActuators(string bluetoothDeviceName, List<int> regions, List<int> positivePins, List<int> negativePins)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.AddActuators(bluetoothDeviceName, regions, positivePins, negativePins));
        }

        public void RemoveActuator(string bluetoothDeviceName, int region)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.RemoveActuator(bluetoothDeviceName, region));
        }

        public void RemoveActuators(string bluetoothDeviceName, List<int> regions)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.RemoveActuators(bluetoothDeviceName, regions));
        }

        public void ActivateActuators(string bluetoothDeviceName, List<int> regions, List<string> intensities)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.ActivateActuators(bluetoothDeviceName, regions, intensities));
        }

        public void ActivateActuatorsTimeTest(string bluetoothDeviceName, List<int> regions, List<string> intensities)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.ActivateActuatorsTimeTest(bluetoothDeviceName, regions, intensities));
        }

        public void TurnOnActuators(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.TurnOnActuators(bluetoothDeviceName));
        }

        public void TurnOffActuators(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.TurnOffActuators(bluetoothDeviceName));
        }

        public void ResetActuators(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.ResetActuators(bluetoothDeviceName));
        }

        public void AddFlexor(string bluetoothDeviceName, int region, int pin)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.AddFlexor(bluetoothDeviceName, region, pin));
        }

        public void AddFlexors(string bluetoothDeviceName, List<int> regions, List<int> pins)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.AddFlexors(bluetoothDeviceName, regions, pins));
        }

        public void RemoveFlexor(string bluetoothDeviceName, int region)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.RemoveFlexor(bluetoothDeviceName, region));
        }

        public void RemoveFlexors(string bluetoothDeviceName, List<int> regions)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.RemoveFlexors(bluetoothDeviceName, regions));
        }

        public void CalibrateFlexors(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.CalibrateFlexors(bluetoothDeviceName));
        }

        public void ConfirmCalibration(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.ConfirmCalibration(bluetoothDeviceName));
        }

        public void SetThreshold(string bluetoothDeviceName, int value)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.SetThreshold(bluetoothDeviceName, value));
        }

        public void TurnOnFlexors(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.TurnOnFlexors(bluetoothDeviceName));
        }

        public void TurnOffFlexors(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.TurnOffFlexors(bluetoothDeviceName));
        }

        public void ResetFlexors(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.ResetFlexors(bluetoothDeviceName));
        }

        public void StartIMU(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.StartIMU(bluetoothDeviceName));
        }

        public void SetIMUStatus(string bluetoothDeviceName, bool status)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.SetIMUStatus(bluetoothDeviceName, status));
        }

        public void SetRawData(string bluetoothDeviceName, bool status)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.SetRawData(bluetoothDeviceName, status));
        }

        public void SetIMUChoosingData(string bluetoothDeviceName, int value)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.SetIMUChoosingData(bluetoothDeviceName, value));
        }

        public void ReadOnlyAccelerometerFromIMU(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.ReadOnlyAccelerometerFromIMU(bluetoothDeviceName));
        }

        public void ReadOnlyGyroscopeFromIMU(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.ReadOnlyGyroscopeFromIMU(bluetoothDeviceName));
        }

        public void ReadOnlyMagnetometerFromIMU(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.ReadOnlyMagnetometerFromIMU(bluetoothDeviceName));
        }

        public void ReadOnlyAttitudeFromIMU(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.ReadOnlyAttitudeFromIMU(bluetoothDeviceName));
        }

        public void ReadAllDataFromIMU(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.ReadAllDataFromIMU(bluetoothDeviceName));
        }

        public void CalibrateIMU(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.CalibrateIMU(bluetoothDeviceName));
        }

        public void TurnOnIMU(string bluetoohDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.TurnOnIMU(bluetoohDeviceName));
        }

        public void TurnOffIMU(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.TurnOffIMU(bluetoothDeviceName));
        }

        public void SetLoopDelay(string bluetoothDeviceName, int value)
        {
            if (this.WebSocket.IsAlive) 
                this.WebSocket.Send(MessageGenerator.SetLoopDelay(bluetoothDeviceName, value));
        }

        public void ConnectToWebSocketServer()
        {
            this.WebSocket.Connect();
            OnWebSocketConnectionStateChangued?.Invoke(this.WebSocket.IsAlive);
        }

        public void DisconnectFromWebSocketServer()
        {
            this.WebSocket.Close();
            OnWebSocketConnectionStateChangued?.Invoke(this.WebSocket.IsAlive);
        }

        public void GetOpenGloveArduinoVersionSoftware(string bluetoothDeviceName)
        {
            if (this.WebSocket.IsAlive)
                this.WebSocket.Send(MessageGenerator.GetOpenGloveArduinoVersionSoftware(bluetoothDeviceName));
        }
    }
}

File: JavaTest/app/src/main/java/com/israel/martinez/javatest/OpenGloveAPI_Java_HL/Communication.java
package com.israel.martinez.javatest.OpenGloveAPI_Java_HL;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.InterfaceAddress;
import java.net.URI;
import java.nio.ByteBuffer;
import java.util.List;

/**
 * Created by israel-martinez on 05-09-18.
 */

public class Communication extends WebSocketClient{

    //public Client WebSocket;
    public URI ServerURI;
    public MessageGenerator MessageGenerator;
    public String BluetoothDeviceName;
    public String ConfigurationName;

    private TimeTestServerLatencyActivateActuatorsReceived OnTimeTestServerLatencyActivateActuatorsReceived  = null;
    private TimeTestArduinoLatencyActivateActuatorsReceived OnTimeTestArduinoLatencyActivateActuatorsReceived = null;
    private FlexorMovement OnFlexorValueReceived  = null;
    private AccelerometerValues OnAccelerometerValuesReceived = null;
    private GyroscopeValues OnGyroscopeValuesReceived = null;
    private MagnetometerValues OnMagnetometerValuesReceived = null;
    private AllIMUValues OnAllIMUValuesReceived = null;
    private BluetoothDeviceConnectionState OnBluetoothDeviceConnectionStateChanged = null;
    private WebSocketConnectionState OnWebSocketConnectionStateChanged = null;
    private InfoMessage OnInfoMessagesReceived = null;

    public Communication(String bluetoothDeviceName, String configurationName, URI serverURI) {
        super(serverURI);
        this.BluetoothDeviceName = bluetoothDeviceName;
        this.ConfigurationName = configurationName;
        this.ServerURI = serverURI;
        this.MessageGenerator = new MessageGenerator(";", ",", "");
    }

    @Override
    public void onOpen(ServerHandshake handshakedata) {
        System.out.println("new connection opened");
        this.send(MessageGenerator.AddOpenGloveDeviceToServer(BluetoothDeviceName, ConfigurationName));
        this.send(MessageGenerator.StartCaptureDataFromServer(BluetoothDeviceName));
        System.out.println("OpenGlove.Communication.AddOpenGloveToServer()");
        System.out.println("Communication.StartCaptureDataFromServer()");

        if (OnWebSocketConnectionStateChanged != null) OnWebSocketConnectionStateChanged.run(this.isOpen());
    }

    @Override
    public void onClose(int code, String reason, boolean remote) {
        System.out.println("closed with exit code " + code + " additional info: " + reason);
        if (OnWebSocketConnectionStateChanged != null) OnWebSocketConnectionStateChanged.run(this.isOpen());
    }

    @Override
    public void onMessage(String message) {
        System.out.println("received message: " + message);
        MessageHandler(message);
    }

    @Override
    public void onMessage(ByteBuffer message) {
        System.out.println("received ByteBuffer");
    }

    @Override
    public void onError(Exception ex) {
        System.err.println("an error occurred:" + ex);
    }


    @FunctionalInterface
    public interface TimeTestServerLatencyActivateActuatorsReceived { void run(long nanoseconds); }
    @FunctionalInterface
    public interface TimeTestArduinoLatencyActivateActuatorsReceived { void run(long nanoseconds); }
    @FunctionalInterface
    public interface FlexorMovement { void run(int mapping, int value); }
    @FunctionalInterface
    public interface AccelerometerValues { void run(float ax, float ay, float az); }
    @FunctionalInterface
    public interface GyroscopeValues { void run(float gx, float gy, float gz); }
    @FunctionalInterface
    public interface MagnetometerValues { void run(float mx, float my, float mz); }
    @FunctionalInterface
    public interface AttitudeValues { void  run(float pitch, float roll, float yaw); }
    @FunctionalInterface
    public interface AllIMUValues { void run(float ax, float ay, float az, float gx, float gy, float gz, float mx, float my, float mz); }
    @FunctionalInterface
    public interface BluetoothDeviceConnectionState { void run(boolean isConnected); }
    @FunctionalInterface
    public interface WebSocketConnectionState { void run(boolean isConnected); }
    @FunctionalInterface
    public interface InfoMessage { void run(String message); }


    public void setOnTimeTestServerLatencyActivateActuatorsReceived(TimeTestServerLatencyActivateActuatorsReceived onTimeTestServerLatencyActivateActuatorsReceived) {
        OnTimeTestServerLatencyActivateActuatorsReceived = onTimeTestServerLatencyActivateActuatorsReceived;
    }

    public void setOnTimeTestArduinoLatencyActivateActuatorsReceived(TimeTestArduinoLatencyActivateActuatorsReceived onTimeTestArduinoLatencyActivateActuatorsReceived) {
        OnTimeTestArduinoLatencyActivateActuatorsReceived = onTimeTestArduinoLatencyActivateActuatorsReceived;
    }

    public void setOnFlexorValueReceived(FlexorMovement onFlexorValueReceived) {
        this.OnFlexorValueReceived = onFlexorValueReceived;
    }

    public void setOnAccelerometerValuesReceived(AccelerometerValues onAccelerometerValuesReceived) {
        OnAccelerometerValuesReceived = onAccelerometerValuesReceived;
    }

    public void setOnGyroscopeValuesReceived(GyroscopeValues onGyroscopeValuesReceived) {
        OnGyroscopeValuesReceived = onGyroscopeValuesReceived;
    }

    public void setOnMagnometerValuesReceived(MagnetometerValues onMagnometerValuesReceived) {
        OnMagnetometerValuesReceived = onMagnometerValuesReceived;
    }

    public void setOnAllIMUValuesReceived(AllIMUValues onAllIMUValuesReceived) {
        OnAllIMUValuesReceived = onAllIMUValuesReceived;
    }

    public void setOnBluetoothDeviceConnectionStateChanged(BluetoothDeviceConnectionState onBluetoothDeviceConnectionStateChanged) {
        OnBluetoothDeviceConnectionStateChanged = onBluetoothDeviceConnectionStateChanged;
    }

    public void setOnWebSocketConnectionStateChanged(WebSocketConnectionState onWebSocketConnectionStateChanged) {
        OnWebSocketConnectionStateChanged = onWebSocketConnectionStateChanged;
    }

    public void setOnInfoMessagesReceived(InfoMessage onInfoMessagesReceived) {
        OnInfoMessagesReceived = onInfoMessagesReceived;
    }

    private void MessageHandler(String message)
    {
        int mapping, value;
        float valueX, valueY, valueZ;
        String[] words;

        if (message != null)
        {
            words = message.split(",");
            try
            {
                switch (words[0])
                {
                    case "f":
                        mapping = Integer.parseInt(words[1]);
                        value = Integer.parseInt(words[2]);
                        if (OnFlexorValueReceived != null)
                            OnFlexorValueReceived.run(mapping, value);
                        break;
                    case "a":
                        valueX = Float.parseFloat(words[1]);
                        valueY = Float.parseFloat(words[2]);
                        valueZ = Float.parseFloat(words[3]);
                        if (OnAccelerometerValuesReceived != null )
                            OnAccelerometerValuesReceived.run(valueX, valueY, valueZ);
                        break;
                    case "g":
                        valueX = Float.parseFloat(words[1]);
                        valueY = Float.parseFloat(words[2]);
                        valueZ = Float.parseFloat(words[3]);
                        if (OnGyroscopeValuesReceived != null)
                            OnGyroscopeValuesReceived.run(valueX, valueY, valueZ);
                        break;
                    case "m":
                        valueX = Float.parseFloat(words[1]);
                        valueY = Float.parseFloat(words[2]);
                        valueZ = Float.parseFloat(words[3]);
                        if (OnMagnetometerValuesReceived != null)
                            OnMagnetometerValuesReceived.run(valueX, valueY, valueZ);
                        break;
                    case "z":
                         if (OnAllIMUValuesReceived != null)
                             OnAllIMUValuesReceived.run(Float.parseFloat(words[1]), Float.parseFloat(words[2]), Float.parseFloat(words[3]), Float.parseFloat(words[4]), Float.parseFloat(words[5]), Float.parseFloat(words[6]), Float.parseFloat(words[7]), Float.parseFloat(words[8]), Float.parseFloat(words[9]));
                        break;
                    case "b":
                        //Received "True" or "False" from WebSocketServer
                        if (OnBluetoothDeviceConnectionStateChanged != null)
                            OnBluetoothDeviceConnectionStateChanged.run(Boolean.parseBoolean(words[1].toLowerCase()));
                        break;
                    case "us":
                        if (OnTimeTestArduinoLatencyActivateActuatorsReceived != null)
                            OnTimeTestArduinoLatencyActivateActuatorsReceived.run(Long.parseLong(words[1]));
                        break;
                    case "ns":
                        if (OnTimeTestServerLatencyActivateActuatorsReceived!= null)
                            OnTimeTestServerLatencyActivateActuatorsReceived.run(Long.parseLong(words[1]));
                        break;

                    default:
                        if (OnInfoMessagesReceived != null)
                            OnInfoMessagesReceived.run(message);
                        break;
                }

            }
            catch (Exception e)
            {
                System.out.println("ERROR: BAD FORMAT on Communication.MessageHandler message: " + message);
            }
        }
    }


    public void StartOpenGlove(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.StartOpenGlove(bluetoothDeviceName));
    }

    public void StopOpenGlove(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.StopOpenGlove(bluetoothDeviceName));
    }

    public void AddOpenGloveDeviceToServer(String bluetoothDeviceName, String configurationName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.AddOpenGloveDeviceToServer(bluetoothDeviceName, configurationName));
    }

    public void RemoveOpenGloveDeviceFromServer(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.RemoveOpenGloveDeviceFromServer(bluetoothDeviceName));
    }
    public void SaveOpenGloveConfiguration(String bluetoothDeviceName, String configurationName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.SaveOpenGloveConfiguration(bluetoothDeviceName, configurationName));
    }

    public void ConnectToBluetoothDevice(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ConnectToBluetoothDevice(bluetoothDeviceName));
    }

    public void DisconnectFromBluetoothDevice(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.DisconnectFromBluetoothDevice(bluetoothDeviceName));
    }

    public void StartCaptureDataFromServer(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.StartCaptureDataFromServer(bluetoothDeviceName));
    }

    public void StopCaptureDataFromServer(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.StopCaptureDataFromServer(bluetoothDeviceName));
    }

    public void AddActuator(String bluetoothDeviceName, int region, int positivePin, int negativePin)
    {
        if (this.isOpen())
            this.send(MessageGenerator.AddActuator(bluetoothDeviceName, region, positivePin, negativePin));
    }

    public void AddActuators(String bluetoothDeviceName, List<Integer> regions, List<Integer> positivePins, List<Integer> negativePins)
    {
        System.out.println("AddActuators: " + MessageGenerator.AddActuators(bluetoothDeviceName, regions, positivePins, negativePins) + "isOpen: "+this.isOpen());
        if (this.isOpen())
            this.send(MessageGenerator.AddActuators(bluetoothDeviceName, regions, positivePins, negativePins));
    }

    public void RemoveActuator(String bluetoothDeviceName, int region)
    {
        if (this.isOpen())
            this.send(MessageGenerator.RemoveActuator(bluetoothDeviceName, region));
    }

    public void RemoveActuators(String bluetoothDeviceName, List<Integer> regions)
    {
        if (this.isOpen())
            this.send(MessageGenerator.RemoveActuators(bluetoothDeviceName, regions));
    }

    public void ActivateActuators(String bluetoothDeviceName, List<Integer> regions, List<String> intensities)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ActivateActuators(bluetoothDeviceName, regions, intensities));
    }

    public void ActivateActuatorsTimeTest(String bluetoothDeviceName, List<Integer> regions, List<String> intensities)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ActivateActuatorsTimeTest(bluetoothDeviceName, regions, intensities));
    }

    public void TurnOnActuators(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.TurnOnActuators(bluetoothDeviceName));
    }

    public void TurnOffActuators(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.TurnOffActuators(bluetoothDeviceName));
    }

    public void ResetActuators(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ResetActuators(bluetoothDeviceName));
    }

    public void AddFlexor(String bluetoothDeviceName, int region, int pin)
    {
        if (this.isOpen())
            this.send(MessageGenerator.AddFlexor(bluetoothDeviceName, region, pin));
    }

    public void AddFlexors(String bluetoothDeviceName, List<Integer> regions, List<Integer> pins)
    {
        System.out.println("AddFlexors: " + MessageGenerator.AddFlexors(bluetoothDeviceName, regions, pins) + "isOpen: "+this.isOpen());
        if (this.isOpen())
            this.send(MessageGenerator.AddFlexors(bluetoothDeviceName, regions, pins));
    }

    public void RemoveFlexor(String bluetoothDeviceName, int region)
    {
        if (this.isOpen())
            this.send(MessageGenerator.RemoveFlexor(bluetoothDeviceName, region));
    }

    public void RemoveFlexors(String bluetoothDeviceName, List<Integer> regions)
    {
        if (this.isOpen())
            this.send(MessageGenerator.RemoveFlexors(bluetoothDeviceName, regions));
    }

    public void CalibrateFlexors(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.CalibrateFlexors(bluetoothDeviceName));
    }

    public void ConfirmCalibration(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ConfirmCalibration(bluetoothDeviceName));
    }

    public void SetThreshold(String bluetoothDeviceName, int value)
    {
        if (this.isOpen())
            this.send(MessageGenerator.SetThreshold(bluetoothDeviceName, value));
    }

    public void TurnOnFlexors(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.TurnOnFlexors(bluetoothDeviceName));
    }

    public void TurnOffFlexors(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.TurnOffFlexors(bluetoothDeviceName));
    }

    public void ResetFlexors(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ResetFlexors(bluetoothDeviceName));
    }

    public void StartIMU(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.StartIMU(bluetoothDeviceName));
    }

    public void SetIMUStatus(String bluetoothDeviceName, boolean status)
    {
        System.out.println("SetIMUStatus: " + MessageGenerator.SetIMUStatus(bluetoothDeviceName, status) + "isOpen: "+this.isOpen());
        if (this.isOpen())
            this.send(MessageGenerator.SetIMUStatus(bluetoothDeviceName, status));
    }

    public void SetRawData(String bluetoothDeviceName, boolean status)
    {
        if (this.isOpen())
            this.send(MessageGenerator.SetRawData(bluetoothDeviceName, status));
    }

    public void SetIMUChoosingData(String bluetoothDeviceName, int value)
    {
        if (this.isOpen())
            this.send(MessageGenerator.SetIMUChoosingData(bluetoothDeviceName, value));
    }

    public void ReadOnlyAccelerometerFromIMU(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ReadOnlyAccelerometerFromIMU(bluetoothDeviceName));
    }

    public void ReadOnlyGyroscopeFromIMU(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ReadOnlyGyroscopeFromIMU(bluetoothDeviceName));
    }

    public void ReadOnlyMagnetometerFromIMU(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ReadOnlyMagnetometerFromIMU(bluetoothDeviceName));
    }

    public void ReadOnlyAttitudeFromIMU(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ReadOnlyAttitudeFromIMU(bluetoothDeviceName));
    }

    public void ReadAllDataFromIMU(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.ReadAllDataFromIMU(bluetoothDeviceName));
    }

    public void CalibrateIMU(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.CalibrateIMU(bluetoothDeviceName));
    }

    public void TurnOnIMU(String bluetoohDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.TurnOnIMU(bluetoohDeviceName));
    }

    public void TurnOffIMU(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.TurnOffIMU(bluetoothDeviceName));
    }

    public void SetLoopDelay(String bluetoothDeviceName, int value)
    {
        if (this.isOpen())
            this.send(MessageGenerator.SetLoopDelay(bluetoothDeviceName, value));
    }
    
    public void ConnectToWebSocketServer()
    {
        this.connect();
        if (OnWebSocketConnectionStateChanged != null) OnWebSocketConnectionStateChanged.run(this.isOpen());
    }

    public void DisconnectFromWebSocketServer()
    {
        this.close();
        if (OnWebSocketConnectionStateChanged != null) OnWebSocketConnectionStateChanged.run(this.isOpen());
    }

    public void GetOpenGloveArduinoVersionSoftware(String bluetoothDeviceName)
    {
        if (this.isOpen())
            this.send(MessageGenerator.GetOpenGloveArduinoVersionSoftware(bluetoothDeviceName));
    }
}
